src/core/stdcpp/array.d

```diff
@@ -0,0 +1,125 @@
+
+extern(C++, std):
+
+/**
+ * D language counterpart to C++ std::array.
+ *
+ * C++ reference: $(LINK2 https://en.cppreference.com/w/cpp/container/array)
+ */
+extern(C++, class) struct array(T, size_t N)
+{
+    ///
+    alias size_type = size_t;
+    ///
+    alias difference_type = ptrdiff_t;
+    ///
+    alias value_type = T;
+    ///
+    alias reference = ref T;
+    ///
+    alias const_reference = ref const(T);
+    ///
+    alias pointer = T*;
+    ///
+    alias const_pointer = const(T)*;
+
+    ///
+    alias as_array this;
+
+    ///
+    extern(D) size_type size() const nothrow @safe @nogc                    { return N; }
+    ///
+    extern(D) size_type max_size() const nothrow @safe @nogc                { return N; }
+    ///
+    extern(D) bool empty() const nothrow @safe @nogc                        { return N == 0; }
+
+    // Element access
+    ///
+    extern(D) reference front() @safe @nogc                                 { static if (N > 0) { return as_array()[0]; } else { return as_array()[][0]; } }
+    ///
+    extern(D) const_reference front() const @safe @nogc                     { static if (N > 0) { return as_array()[0]; } else { return as_array()[][0]; } }
+    ///
+    extern(D) reference back() @safe @nogc                                  { static if (N > 0) { return as_array()[N-1]; } else { return as_array()[][0]; } }
+    ///
+    extern(D) const_reference back() const @safe @nogc                      { static if (N > 0) { return as_array()[N-1]; } else { return as_array()[][0]; } }
+
+    ///
+    extern(D) void fill(ref const(T) value) @safe @nogc                     { foreach (ref T v; as_array()) v = value; }
+
+    // D helpers
+    ///
+    extern(D) size_type opDollar(size_t pos)() const nothrow @safe @nogc    { static assert(pos == 0, "std::vector is one-dimensional"); return N; }
+
+    version(CRuntime_Microsoft)
+    {
+        // perf will be greatly improved by inlining the primitive access functions
+        ///
+        extern(D)        T* data() nothrow @safe @nogc                      { return &_Elems[0]; }
+        ///
+        extern(D) const(T)* data() const nothrow @safe @nogc                { return &_Elems[0]; }
+
+        ///
+        extern(D)        ref T at(size_type i) nothrow @trusted @nogc       { static if (N > 0) { if (N <= i) _Xran(); return _Elems.ptr[i]; } else { _Xran(); return _Elems[0]; } }
+        ///
+        extern(D) ref const(T) at(size_type i) const nothrow @trusted @nogc { static if (N > 0) { if (N <= i) _Xran(); return _Elems.ptr[i]; } else { _Xran(); return _Elems[0]; } }
+
+        ///
+        extern(D) ref inout(T)[N] as_array() const inout @safe @nogc        { return _Elems[0 .. N]; }
+
+    private:
+        import core.stdcpp.xutility : _Xout_of_range;
+
+        T[N ? N : 1] _Elems;
+
+        extern(D) void _Xran() const @safe @nogc                            { _Xout_of_range("invalid array<T, N> subscript"); }
+    }
+    else version(CRuntime_Glibc)
```

Comment:
Nit, you shouldn't be using this version.  C++ bindings are neither CRuntime nor Glibc!

Reply:
Well... yeah, true I guess... what would you suggest?
I mean, libstdc++ is maintained by the same GNU lot as glibc isn't it? Are they separable? I think libstdc++ is just as intrinsically connected to the toolchain as glibc is no?

Reply:
`libstdc++` is part of GCC, but `glibc` is an external project with different maintainers. It is perfectly possible to use `musl` C library instead of `glibc` and link GCC and `libstdc++` against it. Then you have `musl+libstdc++`. Same should be true for other C libraries (dietlibc, uclibc).

Reply:
So, you're saying `libstdc++` is actually closer to the compiler than glibc?
So, should the versions check toolchains rather than cruntimes?
Should we make new versions?

Reply:
> Should we make new versions?

I think so. We also can't use the DMD/LDC/GDC versions as some compilers may use different c++ libraries (I think GCC always uses libstdc++, even on windows. clang however uses libstdc++ for compatibility reasons on linux, but libc++ on OSX and probably the microsoft c++ library for windows). So I think there's no way around new versions.

Reply:
Okay. What's the complete set of likely C++ implementations, and what naming scheme should we roll with?